# Task 6D: SQL vs Pandas
## Shouvanik Sarkar, Student ID: 225174439, email id: s225174439@deakin.edu.au

Importing all the required libraries

In [1]:
import numpy as np
import pandas as pd
pd.set_option("display.notebook_repr_html", False)  # disable "rich" output
import sqlite3

### Reading all the required files for the activity and storing them in the dataframes. 

In [4]:
import os
weather = pd.read_csv(r"D:\Study-Great Learning\Deakin\Data Wrangling\Tasks\data_files\nycflights13_weather\weather.csv", delimiter=',', comment='#')
airlines = pd.read_csv(r"D:\Study-Great Learning\Deakin\Data Wrangling\Tasks\data_files\nycflights13_airlines\airlines.csv" ,delimiter=',', comment='#')
airports = pd.read_csv(r"D:\Study-Great Learning\Deakin\Data Wrangling\Tasks\data_files\nycflights13_airports\airports.csv" ,delimiter=',', comment='#')
flights = pd.read_csv(r"D:\Study-Great Learning\Deakin\Data Wrangling\Tasks\data_files\nycflights13_flights\flights.csv" ,delimiter=',', comment='#')
planes = pd.read_csv(r"D:\Study-Great Learning\Deakin\Data Wrangling\Tasks\data_files\nycflights13_planes\planes.csv" ,delimiter=',', comment='#')

In [ ]:
### Creating a temporary file for Database

creating a temporary data base file in system.

In [6]:
import tempfile, os.path
dbfile = os.path.join(tempfile.mkdtemp(), "travel.db")
print(dbfile)

C:\Users\TARUNK~1\AppData\Local\Temp\tmpk9dtr_54\travel.db


In [8]:
conn = sqlite3.connect(dbfile)##connecting to db file

In [10]:
print(conn)

#### Connecting to Database and reading the dataframes. 

In [12]:
weather.to_sql("weather", conn, index=False)
airlines.to_sql("airlines", conn, index=False)
airports.to_sql("airports", conn, index=False)
flights.to_sql("flights", conn, index=False)
planes.to_sql("planes", conn, index=False)

3322

#### 1st Task
lets check distinct engines in planes table, and lets compare if any differences we observe between dataframe and Structured Query Language. 

In [27]:
task1_s =pd.read_sql_query("""
   SELECT DISTINCT engine FROM planes
""", conn)
print(task1_s)

          engine
0      Turbo-fan
1      Turbo-jet
2  Reciprocating
3        4 Cycle
4    Turbo-shaft
5     Turbo-prop


In [19]:
task1_pd = planes['engine'].unique()
task1_pd = pd.DataFrame(task1_pd)
task1_pd.columns = ['engine']
print(task1_pd)

          engine
0      Turbo-fan
1      Turbo-jet
2  Reciprocating
3        4 Cycle
4    Turbo-shaft
5     Turbo-prop


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.
There are five different types of engine in the dataset for planes. 

Below command will print for any variance between sql and pandas. 

In [30]:
print("Errors for 1st Task(if any):", pd.testing.assert_frame_equal(task1_s,task1_pd))

Errors for 1st Task(if any): None


#### 2nd Task
Lets print for differnt type and engine from the database table planes. And lets print the dataset from both pandas and sql

In [33]:
task2_s =pd.read_sql_query("""
   SELECT DISTINCT type, engine FROM planes
""", conn)
print(task2_s)

                       type         engine
0   Fixed wing multi engine      Turbo-fan
1   Fixed wing multi engine      Turbo-jet
2  Fixed wing single engine  Reciprocating
3   Fixed wing multi engine  Reciprocating
4  Fixed wing single engine        4 Cycle
5                Rotorcraft    Turbo-shaft
6   Fixed wing multi engine     Turbo-prop


In [35]:
task2_pd = planes[['type', 'engine']].drop_duplicates().reset_index(drop = True)
print(task2_pd)

                       type         engine
0   Fixed wing multi engine      Turbo-fan
1   Fixed wing multi engine      Turbo-jet
2  Fixed wing single engine  Reciprocating
3   Fixed wing multi engine  Reciprocating
4  Fixed wing single engine        4 Cycle
5                Rotorcraft    Turbo-shaft
6   Fixed wing multi engine     Turbo-prop


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see there are 6 different type and engine in the dataset. 
And two types of Reciprocating engine in the dataset- Fixed wing single engine and fixed wing multiple engine.  

Below command will print for any variance between sql and pandas statements. 

In [39]:
print("Errors for 2nd Task(if any):", pd.testing.assert_frame_equal(task2_s,task2_pd))

Errors for 2nd Task(if any): None


#### 3rd Task
Lets now check count of differnt types of engines from the dataset planes. 

In [53]:
task3_s =pd.read_sql_query("""
   SELECT COUNT(*), engine FROM planes GROUP BY engine
""", conn)
print(task3_s)

   COUNT(*)         engine
0         2        4 Cycle
1        28  Reciprocating
2      2750      Turbo-fan
3       535      Turbo-jet
4         2     Turbo-prop
5         5    Turbo-shaft


In [55]:
task3_pd = planes.groupby('engine').size().reset_index(name='COUNT(*)')
task3_pd = task3_pd[["COUNT(*)","engine" ]]
print(task3_pd)

   COUNT(*)         engine
0         2        4 Cycle
1        28  Reciprocating
2      2750      Turbo-fan
3       535      Turbo-jet
4         2     Turbo-prop
5         5    Turbo-shaft


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see there are 5 different types of engine in the dataset. 
with most of the type being Turbo Fan type of Engine. Almost 5 times compared to sum total of other types. 

Below command will print for any variance between sql and pandas statements. 

In [58]:
print("Errors for 3rd Task(if any):", pd.testing.assert_frame_equal(task3_s,task3_pd))

Errors for 3rd Task(if any): None


#### 4th Task
Lets now check count of engine and different types in plane database. 

In [61]:
task4_s =pd.read_sql_query("""
   SELECT COUNT(*), engine, type FROM planes GROUP BY engine, type
""", conn)
print(task4_s)

   COUNT(*)         engine                      type
0         2        4 Cycle  Fixed wing single engine
1         5  Reciprocating   Fixed wing multi engine
2        23  Reciprocating  Fixed wing single engine
3      2750      Turbo-fan   Fixed wing multi engine
4       535      Turbo-jet   Fixed wing multi engine
5         2     Turbo-prop   Fixed wing multi engine
6         5    Turbo-shaft                Rotorcraft


In [63]:
task4_pd = planes.groupby(['engine', 'type']).size().reset_index(name='COUNT(*)')
task4_pd = task4_pd[["COUNT(*)","engine","type" ]]
print(task4_pd)

   COUNT(*)         engine                      type
0         2        4 Cycle  Fixed wing single engine
1         5  Reciprocating   Fixed wing multi engine
2        23  Reciprocating  Fixed wing single engine
3      2750      Turbo-fan   Fixed wing multi engine
4       535      Turbo-jet   Fixed wing multi engine
5         2     Turbo-prop   Fixed wing multi engine
6         5    Turbo-shaft                Rotorcraft


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see there are 6 different type and engine in the dataset, as we saw before.  
And two types of Reciprocating engine in the dataset- Fixed wing single engine are more compared to fixed wing multiple engine.  

Below command will print for any variance between sql and pandas statements. 

In [66]:
print("Errors for 4th Task(if any):", pd.testing.assert_frame_equal(task4_s,task4_pd))

Errors for 4th Task(if any): None


#### 5th Task
Lets now check minimum, maximum and average of the year the engines were manufactured along with different manufactures we have

In [73]:
task5_s =pd.read_sql_query("""
   SELECT MIN(year), AVG(year), MAX(year), engine, manufacturer
   FROM planes
   GROUP BY engine, manufacturer
""", conn)
print(task5_s)

    MIN(year)    AVG(year)  MAX(year)         engine  \
0      1975.0  1975.000000     1975.0        4 Cycle   
1         NaN          NaN        NaN        4 Cycle   
2         NaN          NaN        NaN  Reciprocating   
3      2007.0  2007.000000     2007.0  Reciprocating   
4         NaN          NaN        NaN  Reciprocating   
5      1959.0  1971.142857     1983.0  Reciprocating   
6      2007.0  2007.000000     2007.0  Reciprocating   
7      1959.0  1959.000000     1959.0  Reciprocating   
8      1956.0  1956.000000     1956.0  Reciprocating   
9      2007.0  2007.000000     2007.0  Reciprocating   
10        NaN          NaN        NaN  Reciprocating   
11     1985.0  1985.000000     1985.0  Reciprocating   
12        NaN          NaN        NaN  Reciprocating   
13     1985.0  1985.000000     1985.0  Reciprocating   
14     1993.0  1993.000000     1993.0  Reciprocating   
15        NaN          NaN        NaN  Reciprocating   
16     1968.0  1976.400000     1980.0  Reciproca

In [83]:
task5_pd = planes[["engine","manufacturer","year"]].groupby(["engine","manufacturer"])["year"].agg([np.min,np.mean,np.max])\
          .sort_values(["engine","manufacturer"],ascending = [True,True])\
          .reset_index()
task5_pd = task5_pd[["min","mean","max", "engine", "manufacturer"]]
task5_pd.columns=["MIN(year)",  "AVG(year)", "MAX(year)",  "engine", "manufacturer"]
print(task5_pd)

    MIN(year)    AVG(year)  MAX(year)         engine  \
0      1975.0  1975.000000     1975.0        4 Cycle   
1         NaN          NaN        NaN        4 Cycle   
2         NaN          NaN        NaN  Reciprocating   
3      2007.0  2007.000000     2007.0  Reciprocating   
4         NaN          NaN        NaN  Reciprocating   
5      1959.0  1971.142857     1983.0  Reciprocating   
6      2007.0  2007.000000     2007.0  Reciprocating   
7      1959.0  1959.000000     1959.0  Reciprocating   
8      1956.0  1956.000000     1956.0  Reciprocating   
9      2007.0  2007.000000     2007.0  Reciprocating   
10        NaN          NaN        NaN  Reciprocating   
11     1985.0  1985.000000     1985.0  Reciprocating   
12        NaN          NaN        NaN  Reciprocating   
13     1985.0  1985.000000     1985.0  Reciprocating   
14     1993.0  1993.000000     1993.0  Reciprocating   
15        NaN          NaN        NaN  Reciprocating   
16     1968.0  1976.400000     1980.0  Reciproca

C:\Users\TARUN KUMAR SARKAR\AppData\Local\Temp\ipykernel_14756\2344041619.py:1: FutureWarning: The provided callable <function min at 0x000001FF60B7CAE0> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  task5_pd = planes[["engine","manufacturer","year"]].groupby(["engine","manufacturer"])["year"].agg([np.min,np.mean,np.max])\
C:\Users\TARUN KUMAR SARKAR\AppData\Local\Temp\ipykernel_14756\2344041619.py:1: FutureWarning: The provided callable <function mean at 0x000001FF60B7D3A0> is currently using SeriesGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  task5_pd = planes[["engine","manufacturer","year"]].groupby(["engine","manufacturer"])["year"].agg([np.min,np.mean,np.max])\
C:\Users\TARUN KUMAR SARKAR\AppData\Local\Temp\ipykernel_14756\2344041619.py:1: FutureWarning: 

#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see for few manufacturers, the year of manufacture is not present for which we are getting error as NaN. Oldest manufacturer is of 1967 with manufatcurer as Beech.

Below command will print for any variance between sql and pandas statements. 

In [85]:
print("Errors for 5th Task(if any):", pd.testing.assert_frame_equal(task5_s,task5_pd))

Errors for 5th Task(if any): None


#### 6th Task
Lets now pull the data from planes where speed is present in the dataset. we will not pull records where speed is not present. 

In [89]:
task6_s =pd.read_sql_query("""
   SELECT * FROM planes WHERE speed IS NOT NULL
""", conn)
print(task6_s)

   tailnum    year                      type       manufacturer        model  \
0   N201AA  1959.0  Fixed wing single engine             CESSNA          150   
1   N202AA  1980.0   Fixed wing multi engine             CESSNA         421C   
2   N350AA  1980.0   Fixed wing multi engine              PIPER    PA-31-350   
3   N364AA  1973.0   Fixed wing multi engine             CESSNA         310Q   
4   N378AA  1963.0  Fixed wing single engine             CESSNA         172E   
5   N381AA  1956.0   Fixed wing multi engine            DOUGLAS       DC-7BF   
6   N425AA  1968.0  Fixed wing single engine              PIPER    PA-28-180   
7   N508AA  1975.0                Rotorcraft               BELL         206B   
8   N519MQ  1983.0  Fixed wing single engine             CESSNA        A185F   
9   N525AA  1980.0   Fixed wing multi engine              PIPER    PA-31-350   
10  N545AA  1976.0  Fixed wing single engine              PIPER   PA-32R-300   
11  N567AA  1959.0  Fixed wing single en

In [91]:
task6_pd = planes.loc[(~planes.speed.isna())].reset_index(drop = True)
print(task6_pd)

   tailnum    year                      type       manufacturer        model  \
0   N201AA  1959.0  Fixed wing single engine             CESSNA          150   
1   N202AA  1980.0   Fixed wing multi engine             CESSNA         421C   
2   N350AA  1980.0   Fixed wing multi engine              PIPER    PA-31-350   
3   N364AA  1973.0   Fixed wing multi engine             CESSNA         310Q   
4   N378AA  1963.0  Fixed wing single engine             CESSNA         172E   
5   N381AA  1956.0   Fixed wing multi engine            DOUGLAS       DC-7BF   
6   N425AA  1968.0  Fixed wing single engine              PIPER    PA-28-180   
7   N508AA  1975.0                Rotorcraft               BELL         206B   
8   N519MQ  1983.0  Fixed wing single engine             CESSNA        A185F   
9   N525AA  1980.0   Fixed wing multi engine              PIPER    PA-31-350   
10  N545AA  1976.0  Fixed wing single engine              PIPER   PA-32R-300   
11  N567AA  1959.0  Fixed wing single en

#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see there are 22 different records for which speed is present and printed. 

Below command will print for any variance between sql and pandas statements. 

In [94]:
print("Errors for 6th Task(if any):", pd.testing.assert_frame_equal(task6_s,task6_pd))

Errors for 6th Task(if any): None


#### 7th Task
Lets now print the tail number for all the records for which the seats are between 150 and 210 and year is greater than 2011

In [96]:
task7_s =pd.read_sql_query("""
   SELECT tailnum FROM planes
WHERE seats BETWEEN 150 AND 210 AND year >= 2011
""", conn)
print(task7_s)

   tailnum
0   N150UW
1   N151UW
2   N152UW
3   N153UW
4   N154UW
..     ...
87  N851VA
88  N852VA
89  N853VA
90  N854VA
91  N855VA

[92 rows x 1 columns]


In [99]:
task7_pd = planes.loc[(planes.seats >=150) & (planes.seats < 211) & (planes.year >=2011),["tailnum"]].reset_index(drop = True)
print(task7_pd)

   tailnum
0   N150UW
1   N151UW
2   N152UW
3   N153UW
4   N154UW
..     ...
87  N851VA
88  N852VA
89  N853VA
90  N854VA
91  N855VA

[92 rows x 1 columns]


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We have printed 92 different names of tail numbers as present in the dataset planes. 

Below command will print for any variance between sql and pandas statements. 

In [102]:
print("Errors for 7th Task(if any):", pd.testing.assert_frame_equal(task7_s,task7_pd))

Errors for 7th Task(if any): None


#### 8th Task
Lets now check the tail numbers, manufacturer ad seats from the same dataset planes where manufacturer are BOEING or AIRBUS or EMBRAER and seats are more than 390

In [105]:
task8_s =pd.read_sql_query("""
   SELECT tailnum, manufacturer, seats FROM planes
WHERE manufacturer IN ("BOEING", "AIRBUS", "EMBRAER") AND seats>390
""", conn)
print(task8_s)

   tailnum manufacturer  seats
0   N206UA       BOEING    400
1   N228UA       BOEING    400
2   N272AT       BOEING    400
3   N57016       BOEING    400
4   N670US       BOEING    450
5   N77012       BOEING    400
6   N777UA       BOEING    400
7   N78003       BOEING    400
8   N78013       BOEING    400
9   N787UA       BOEING    400
10  N862DA       BOEING    400
11  N863DA       BOEING    400
12  N865DA       BOEING    400


In [107]:
task8_pd = planes.loc[(planes.manufacturer.isin(['BOEING', 'AIRBUS', 'EMBRAER'])) & (planes.seats > 390),['tailnum', 'manufacturer','seats']].reset_index(drop = True)
print(task8_pd)

   tailnum manufacturer  seats
0   N206UA       BOEING    400
1   N228UA       BOEING    400
2   N272AT       BOEING    400
3   N57016       BOEING    400
4   N670US       BOEING    450
5   N77012       BOEING    400
6   N777UA       BOEING    400
7   N78003       BOEING    400
8   N78013       BOEING    400
9   N787UA       BOEING    400
10  N862DA       BOEING    400
11  N863DA       BOEING    400
12  N865DA       BOEING    400


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see there are 12 different tail numbers all for manufacturer BOEING amongst AIRBUS and EMBRAER for seats more than 390. 

Below command will print for any variance between sql and pandas statements. 

In [110]:
print("Errors for 8th Task(if any):", pd.testing.assert_frame_equal(task8_s,task8_pd))

Errors for 8th Task(if any): None


#### 9th Task
Lets now print differnt year and seats from the dataset planes where year is more than 2012. 

In [113]:
task9_s =pd.read_sql_query("""
   SELECT DISTINCT year, seats FROM planes
WHERE year >= 2012 ORDER BY year ASC, seats DESC
""", conn)
print(task9_s)

      year  seats
0   2012.0    379
1   2012.0    377
2   2012.0    260
3   2012.0    222
4   2012.0    200
5   2012.0    191
6   2012.0    182
7   2012.0    149
8   2012.0    140
9   2012.0     20
10  2012.0      5
11  2013.0    379
12  2013.0    377
13  2013.0    222
14  2013.0    200
15  2013.0    199
16  2013.0    191
17  2013.0    182
18  2013.0    140
19  2013.0     95
20  2013.0     20


In [115]:
task9_pd = planes.loc[(planes.year >= 2012),['year', 'seats']].drop_duplicates().sort_values(["year","seats"],ascending = [True,False]).reset_index(drop = True)
print(task9_pd)

      year  seats
0   2012.0    379
1   2012.0    377
2   2012.0    260
3   2012.0    222
4   2012.0    200
5   2012.0    191
6   2012.0    182
7   2012.0    149
8   2012.0    140
9   2012.0     20
10  2012.0      5
11  2013.0    379
12  2013.0    377
13  2013.0    222
14  2013.0    200
15  2013.0    199
16  2013.0    191
17  2013.0    182
18  2013.0    140
19  2013.0     95
20  2013.0     20


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see there are 20 different types of seats and year. with half of them for the year 2012 and rest half in 2013. 

Below command will print for any variance between sql and pandas statements. 

In [118]:
print("Errors for 9th Task(if any):", pd.testing.assert_frame_equal(task9_s,task9_pd))

Errors for 9th Task(if any): None


#### 10th Task
Lets now print the same requirement but sorted by most number of seats and least from years. 

In [121]:
task10_s =pd.read_sql_query("""
   SELECT DISTINCT year, seats FROM planes
WHERE year >= 2012 ORDER BY seats DESC, year ASC
""", conn)
print(task10_s)

      year  seats
0   2012.0    379
1   2013.0    379
2   2012.0    377
3   2013.0    377
4   2012.0    260
5   2012.0    222
6   2013.0    222
7   2012.0    200
8   2013.0    200
9   2013.0    199
10  2012.0    191
11  2013.0    191
12  2012.0    182
13  2013.0    182
14  2012.0    149
15  2012.0    140
16  2013.0    140
17  2013.0     95
18  2012.0     20
19  2013.0     20
20  2012.0      5


In [125]:
task10_pd = planes.loc[(planes.year >= 2012),['year', 'seats']].drop_duplicates().sort_values(["seats","year"],ascending = [False,True]).reset_index(drop = True)
print(task10_pd)

      year  seats
0   2012.0    379
1   2013.0    379
2   2012.0    377
3   2013.0    377
4   2012.0    260
5   2012.0    222
6   2013.0    222
7   2012.0    200
8   2013.0    200
9   2013.0    199
10  2012.0    191
11  2013.0    191
12  2012.0    182
13  2013.0    182
14  2012.0    149
15  2012.0    140
16  2013.0    140
17  2013.0     95
18  2012.0     20
19  2013.0     20
20  2012.0      5


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

As we saw there are 20 different types of seats and year. with half of them for the year 2012 and rest half in 2013. 
Most number of seats of 379 and 377 were manufactured in both 2012 and 2013. 
But the least number of seats of 5 were only manufactured in the year 2012. 

Below command will print for any variance between sql and pandas statements. 

In [129]:
print("Errors for 10th Task(if any):", pd.testing.assert_frame_equal(task10_s,task10_pd))

Errors for 10th Task(if any): None


#### 11th Task
Lets now print the manufactures and count of planes manufatured for seats more than 200. 

In [132]:
task11_s =pd.read_sql_query("""
   SELECT manufacturer, COUNT(*) FROM planes
   WHERE seats > 200 GROUP BY manufacturer
""", conn)
print(task11_s)

       manufacturer  COUNT(*)
0            AIRBUS        66
1  AIRBUS INDUSTRIE         4
2            BOEING       225


In [134]:
task11_pd = planes.loc[(planes.seats > 200)].groupby(['manufacturer']).size().reset_index(name='COUNT(*)')
print(task11_pd)

       manufacturer  COUNT(*)
0            AIRBUS        66
1  AIRBUS INDUSTRIE         4
2            BOEING       225


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

Most number of Planes were manufactured by Boeing 225, then by Airbus of 66 units followed by Airbus Industried of 4. 

Below command will print for any variance between sql and pandas statements. 

In [137]:
print("Errors for 11th Task(if any):", pd.testing.assert_frame_equal(task11_s,task11_pd))

Errors for 11th Task(if any): None


#### 12th Task
Lets now print manufactures and count of planes manufatured with count of planes more than 10.

In [139]:
task12_s =pd.read_sql_query("""
   SELECT manufacturer, COUNT(*) FROM planes
   GROUP BY manufacturer HAVING COUNT(*) > 10
""", conn)
print(task12_s)

                    manufacturer  COUNT(*)
0                         AIRBUS       336
1               AIRBUS INDUSTRIE       400
2                         BOEING      1630
3                 BOMBARDIER INC       368
4                        EMBRAER       299
5              MCDONNELL DOUGLAS       120
6  MCDONNELL DOUGLAS AIRCRAFT CO       103
7  MCDONNELL DOUGLAS CORPORATION        14


In [142]:
task12_pd = planes.groupby(['manufacturer']).size().reset_index(name='COUNT(*)').query('`COUNT(*)` > 10').reset_index(drop = True)
print(task12_pd)

                    manufacturer  COUNT(*)
0                         AIRBUS       336
1               AIRBUS INDUSTRIE       400
2                         BOEING      1630
3                 BOMBARDIER INC       368
4                        EMBRAER       299
5              MCDONNELL DOUGLAS       120
6  MCDONNELL DOUGLAS AIRCRAFT CO       103
7  MCDONNELL DOUGLAS CORPORATION        14


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

Most number of Planes were manufactured by Boeing 1630, then by Airbus Industries amd Bombardier inc.

Below command will print for any variance between sql and pandas statements.

In [145]:
print("Errors for 12th Task(if any):", pd.testing.assert_frame_equal(task12_s,task12_pd))

Errors for 12th Task(if any): None


#### 13th Task
Lets now print the manufactures and count of planes manufatured for seats more than 200 with count of manufactured more than 10. 

In [152]:
task13_s =pd.read_sql_query("""
   SELECT manufacturer, COUNT(*) FROM planes
WHERE seats > 200 GROUP BY manufacturer HAVING COUNT(*) > 10
""", conn)
print(task13_s)

  manufacturer  COUNT(*)
0       AIRBUS        66
1       BOEING       225


In [154]:
task13_pd = planes.loc[(planes.seats > 200)].groupby(['manufacturer']).size().reset_index(name='COUNT(*)').query('`COUNT(*)` > 10').reset_index(drop = True)
print(task13_pd)

  manufacturer  COUNT(*)
0       AIRBUS        66
1       BOEING       225


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

As we saw before, only two records remain and AIRBUS Industries falls off the list as it manufactured only 4. 

Below command will print for any variance between sql and pandas statements.

In [158]:
print("Errors for 13th Task(if any):", pd.testing.assert_frame_equal(task13_s,task13_pd))

Errors for 13th Task(if any): None


#### 14th Task
Lets now print manufacturer and how many planes were manufactured for the most manufactured firs 10 records. 

In [161]:
task14_s =pd.read_sql_query("""
   SELECT manufacturer, COUNT(*) AS howmany
FROM planes
GROUP BY manufacturer
ORDER BY howmany DESC LIMIT 10
""", conn)
print(task14_s)

                    manufacturer  howmany
0                         BOEING     1630
1               AIRBUS INDUSTRIE      400
2                 BOMBARDIER INC      368
3                         AIRBUS      336
4                        EMBRAER      299
5              MCDONNELL DOUGLAS      120
6  MCDONNELL DOUGLAS AIRCRAFT CO      103
7  MCDONNELL DOUGLAS CORPORATION       14
8                         CESSNA        9
9                       CANADAIR        9


In [163]:
task14_pd = planes.groupby(['manufacturer']).size().reset_index(name='howmany').reset_index(drop = True).sort_values('howmany',ascending = False).head(10).reset_index(drop=True)
print(task14_pd)

                    manufacturer  howmany
0                         BOEING     1630
1               AIRBUS INDUSTRIE      400
2                 BOMBARDIER INC      368
3                         AIRBUS      336
4                        EMBRAER      299
5              MCDONNELL DOUGLAS      120
6  MCDONNELL DOUGLAS AIRCRAFT CO      103
7  MCDONNELL DOUGLAS CORPORATION       14
8                         CESSNA        9
9                       CANADAIR        9


#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

As we saw before, most planes were manufactured by BOEING followed by AIRBUS Indistries, so on and so forth. 

Below command will print for any variance between sql and pandas statements.

In [166]:
print("Errors for 14th Task(if any):", pd.testing.assert_frame_equal(task14_s,task14_pd))

Errors for 14th Task(if any): None


#### 15th Task
lets now print all the details of dataset flights and year of manufacture, speed and number of seats in the planes. 

In [170]:
task15_s =pd.read_sql_query("""
   SELECT flights.*, planes.year AS plane_year, planes.speed AS plane_speed,
planes.seats AS plane_seats
FROM flights LEFT JOIN planes ON flights.tailnum=planes.tailnum
""", conn)
print(task15_s)

        year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  ...  origin dest air_time distance  \
0                 

In [172]:
task15_pd = pd.merge(flights, planes[['tailnum', 'year', 'speed', 'seats']],
    how='left',on='tailnum'
).rename(columns={'year_x': 'year','year_y' : 'plane_year', 'speed' : 'plane_speed','seats':'plane_seats' })
print(task15_pd)

        year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  ...  origin dest air_time distance  \
0                 

#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

Below command will print for any variance between sql and pandas statements.

In [ ]:
print("Errors for 15th Task(if any):", pd.testing.assert_frame_equal(task15_s,task15_pd))

#### 16th Task
Lets now print all different features of planes and airplanes while getting details of carrier and tail numbers. 

In [176]:
task16_s =pd.read_sql_query("""
   SELECT planes.*, airlines.* FROM
(SELECT DISTINCT carrier, tailnum FROM flights) AS cartail
INNER JOIN planes ON cartail.tailnum=planes.tailnum
INNER JOIN airlines ON cartail.carrier=airlines.carrier
""", conn)
task16_s=task16_s.sort_values(['tailnum','year','type','manufacturer','model','engines','seats','speed','engine','carrier','name'])
print(task16_s)

     tailnum    year                     type                   manufacturer  \
0     N10156  2004.0  Fixed wing multi engine                        EMBRAER   
1     N102UW  1998.0  Fixed wing multi engine               AIRBUS INDUSTRIE   
2     N103US  1999.0  Fixed wing multi engine               AIRBUS INDUSTRIE   
3     N104UW  1999.0  Fixed wing multi engine               AIRBUS INDUSTRIE   
4     N10575  2002.0  Fixed wing multi engine                        EMBRAER   
...      ...     ...                      ...                            ...   
3334  N997AT  2002.0  Fixed wing multi engine                         BOEING   
3335  N997DL  1992.0  Fixed wing multi engine  MCDONNELL DOUGLAS AIRCRAFT CO   
3336  N998AT  2002.0  Fixed wing multi engine                         BOEING   
3337  N998DL  1992.0  Fixed wing multi engine  MCDONNELL DOUGLAS CORPORATION   
3338  N999DN  1992.0  Fixed wing multi engine  MCDONNELL DOUGLAS CORPORATION   

          model  engines  seats  speed 

In [178]:
cartail = flights[['carrier', 'tailnum']].drop_duplicates()
task16_pd =pd.merge(planes,cartail, how = 'inner', on='tailnum')
task16_pd =pd.merge(task16_pd,airlines, how = 'inner', on='carrier')
task16_pd =task16_pd.sort_values(['tailnum','year','type','manufacturer','model','engines','seats','speed','engine','carrier','name']).reset_index(drop=True)
print(task16_pd)

     tailnum    year                     type                   manufacturer  \
0     N10156  2004.0  Fixed wing multi engine                        EMBRAER   
1     N102UW  1998.0  Fixed wing multi engine               AIRBUS INDUSTRIE   
2     N103US  1999.0  Fixed wing multi engine               AIRBUS INDUSTRIE   
3     N104UW  1999.0  Fixed wing multi engine               AIRBUS INDUSTRIE   
4     N10575  2002.0  Fixed wing multi engine                        EMBRAER   
...      ...     ...                      ...                            ...   
3334  N997AT  2002.0  Fixed wing multi engine                         BOEING   
3335  N997DL  1992.0  Fixed wing multi engine  MCDONNELL DOUGLAS AIRCRAFT CO   
3336  N998AT  2002.0  Fixed wing multi engine                         BOEING   
3337  N998DL  1992.0  Fixed wing multi engine  MCDONNELL DOUGLAS CORPORATION   
3338  N999DN  1992.0  Fixed wing multi engine  MCDONNELL DOUGLAS CORPORATION   

          model  engines  seats  speed 

#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

Below command will print for any variance between sql and pandas statements.

In [181]:
print("Errors for 16th Task(if any):", pd.testing.assert_frame_equal(task16_s,task16_pd))

Errors for 16th Task(if any): None


#### 17th Task
Lets now check for all the flights which were originated from EWR and check for the average temperature and humidity throughout the year. 

In [184]:
task17_s =pd.read_sql_query("""
   SELECT flights2.*, atemp, ahumid
FROM (
SELECT * FROM flights WHERE origin='EWR'
) AS flights2 LEFT JOIN (
SELECT year, month, day, AVG(temp) AS atemp,
AVG(humid) AS ahumid
FROM weather
WHERE origin='EWR'
GROUP BY year, month, day ) AS weather2
ON flights2.year=weather2.year
AND flights2.month=weather2.month
AND flights2.day=weather2.day
""", conn)
print(task17_s)

        year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     554.0             558       -4.0     740.0   
2       2013      1    1     555.0             600       -5.0     913.0   
3       2013      1    1     558.0             600       -2.0     923.0   
4       2013      1    1     559.0             600       -1.0     854.0   
...      ...    ...  ...       ...             ...        ...       ...   
120830  2013      9   30    2142.0            2129       13.0    2250.0   
120831  2013      9   30    2149.0            2156       -7.0    2245.0   
120832  2013      9   30    2150.0            2159       -9.0    2250.0   
120833  2013      9   30    2211.0            2059       72.0    2339.0   
120834  2013      9   30    2233.0            2113       80.0     112.0   

        sched_arr_time  arr_delay carrier  ...  tailnum origin dest air_time  \
0                  

In [186]:
weather2 = weather.loc[(weather.origin=='EWR')].groupby(["year", "month", "day"])[["temp", "humid"]].mean()
weather2.columns=["atemp", "ahumid"]
weather2=weather2.reset_index()
flights2 =flights.loc[(flights.origin == 'EWR')].reset_index(drop = True)
task17_pd=pd.merge(flights2,weather2,how = 'left', on=['year','month','day'])
print(task17_pd)

        year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     554.0             558       -4.0     740.0   
2       2013      1    1     555.0             600       -5.0     913.0   
3       2013      1    1     558.0             600       -2.0     923.0   
4       2013      1    1     559.0             600       -1.0     854.0   
...      ...    ...  ...       ...             ...        ...       ...   
120830  2013      9   30    2142.0            2129       13.0    2250.0   
120831  2013      9   30    2149.0            2156       -7.0    2245.0   
120832  2013      9   30    2150.0            2159       -9.0    2250.0   
120833  2013      9   30    2211.0            2059       72.0    2339.0   
120834  2013      9   30    2233.0            2113       80.0     112.0   

        sched_arr_time  arr_delay carrier  ...  tailnum origin dest air_time  \
0                  

#### Compare the result set of pandas and database
On printing both from sql and pandas, we can see values match exactly.

We can see all differnt departure times and delay in departure or arrival time along with distance covered. 

Below command will print for any variance between sql and pandas statements.

In [189]:
print("Errors for 17th Task(if any):", pd.testing.assert_frame_equal(task17_s,task17_pd))

C:\Users\TARUN KUMAR SARKAR\AppData\Local\Temp\ipykernel_14756\2519020596.py:1: FutureWarning: Mismatched null-like values None and nan found. In a future version, pandas equality-testing functions (e.g. assert_frame_equal) will consider these not-matching and raise.
  print("Errors for 17th Task(if any):", pd.testing.assert_frame_equal(task17_s,task17_pd))


Errors for 17th Task(if any): None


# Conclusion
### We were able to check and compare for all the results sets for queries and equivalent pandas code set return values which matched.
We compared various results set and can see sql and pandas returning similar data set. 
Although we might need to do further EDA to understand the run times and various departure and arrival nodes. 
We did check various types of planes and manufactures and various types of engines in the initial set of tasks. 